module/zfs/zfs_log.c

```diff
@@ -525,6 +526,125 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype, znode_t *sdzp,
 	zil_itx_assign(zilog, itx, tx);
 }
 
+/*
+ * At the moment, only Linux supports the renameat2 variant of renameat, which
+ * adds three new flags of interest for us:
+ *
+ *     RENAME_NOREPLACE: if the target name at the moment of the call exists,
+ *                       don't rewrite it and return error
+ *     RENAME_EXCHANGE: atomically swap the two names on the filesystem
+ *     RENAME_WHITEOUT: creates a whiteout inode in place of renamed file as
+ *                      an atomic operation
+ *
+ * Ideally, these operations should be represented as new ZFS Intent Log
+ * txtypes, which would mandate a new ZFS feature flag due to the on-disk
+ * format change. One would then use spa_feature_incr/decr functions to
+ * indicate that the on-disk log contains these new txtypes. However, these
+ * functions are only supposed to be called from the txg syncing context.
+ *
+ * This means that we would need to force out an in-progress txg to disk and
+ * start a new one before writing any ZIL records. This would ensure that
+ * previous versions of ZFS which do not support these log txtypes would never
+ * encounter them during ZIL replay. Doing this would hurt performance.
```

Comment:
That's an interesting idea.  If we cleared the feature flag on a clean unmount, then all we'd be requiring to preserve portability is that the filesystems be unmounted before the pool is moved between incompatible versions.  That's not bad at all, I think this would be pretty reasonable.